src/ocean/util/app/DaemonApp.d

```diff
@@ -549,19 +553,32 @@ public abstract class DaemonApp : Application,
 
     /***************************************************************************
 
-        ISignalExtExtension methods dummy implementation.
+        ISignalExtExtension method default implementation.
 
-        This method is implemented with an "empty" implementation to ease
-        deriving from this class.
+        This method is implemented with behaviour most commonly desired in apps
+        that don't do any custom signal handling - attempt cleaner shutdown on
+        SIGTERM signal.
 
         See ISignalExtExtension documentation for more information on how to
-        override this method.
+        override this method with own behaviour. `super.onSignal` is not needed
+        to be called when doing so.
 
     ***************************************************************************/
 
     override public void onSignal ( int signum )
     {
-        // Dummy implementation of the interface
+        switch ( signum )
+        {
+            case core.sys.posix.signal.SIGTERM:
+                // Default implementation to shut down cleanly
+                if (isSchedulerUsed())
+                    theScheduler.shutdown();
+                else
+                    exit(signum);
```

Comment:
I think this is the only loose end in this PR. This is not semantically same thing as removing the installed signal handler for this and killing the program with the same signal. For example, in `systemd`, with `Restart=on-failure` this will cause service to restart (exit with the non-zero code), while in reality it shouldn't (as `SIGTERM` is considered a "clean" signal by default). I would extend `SignalExt` by making portion of this code `public` (https://github.com/sociomantic-tsunami/ocean/blob/v3.x.x/src/ocean/util/app/ext/SignalExt.d#L435) where it should just accept a signal number to restore to the previous handler (akin to `ignoreSignal`) and then call that and call `kill` again.

It's a bit more work, but I don't think it's too hard to get this working.

Reply:
> SIGTERM is considered a "clean" signal by default)

This doesn't match my observations at all. Simple test with `killall -SIGTERM yes` shows exit status of 143.

Reply:
But that's just shell convention, and the shell is getting this value from the `wait (2)`, inspecting:

```
WIFSIGNALED(stat_val)
Evaluates to a non-zero value if status was returned for a child process that terminated due to the receipt of a signal that was not caught (see <signal.h>).
WTERMSIG(stat_val)
If the value of WIFSIGNALED(stat_val) is non-zero, this macro evaluates to the number of the signal that caused the termination of the child process.
```

so it converts this to an exit return code. In case of the init system, there's no shell to do such thing, and instead of not restarting daemon (on `restart=on-failure` setting of systemd), it will figure out non-zero exit code (inspecting `WEXITED`/`WEXITSTATUS`) and it will restart the program, which is wrong.

Reply:
Here's how bash calculates the exit code: https://github.com/tpruzina/bash/blob/35751626eebdec08c6ac8600b33477e6db31988a/nojobs.c#L216-L224

and here's how it's important to systemd if the service exited via `exit` or via unhandled signal:

https://github.com/systemd/systemd-stable/blob/dbd3dedfaf9165c5b9a0f159ac26c2a2bec2dae1/src/shared/exit-status.c#L165-L180

(using `waitid`:

```
       si_code     Set to one of: CLD_EXITED (child called _exit(2)); CLD_KILLED (child killed by signal); CLD_DUMPED (child killed by signal, and dumped core); CLD_STOPPED (child stopped  by  sig‐
                   nal); CLD_TRAPPED (traced child has trapped); or CLD_CONTINUED (child continued by SIGCONT).
```)

Reply:
Thinking about this again - is there any reason why we can't simply do `epoll.shutdown()` for the second branch? It is not as reliable as scheduler version but worst case SIGKILL will just arrive next :)

Reply:
Huh, I assumed there's no `EpollSelectDispatcher` available in that context. I might be wrong.

Reply:
Aha, there is, it gets assigned to the private `this.epoll` instance here: https://github.com/sociomantic-tsunami/ocean/blob/v3.x.x/src/ocean/util/app/DaemonApp.d#L406

In that case, you don't even need a branch for the scheduler: `DaemonApp.epoll` will point to a scheduler's one.

Reply:
`theScheduler.shutdown` does a bit more than just calling epoll one - it also kills all tasks and prevents new ones from being spawned

Reply:
Ah, fair enough. I think `this.epoll.shutdown()` in another branch is a fair compromise.